File: packages/fble/android/src/main/java/io/flutter/plugins/fble/FblePlugin.java
// Copyright 2017, the Flutter project authors. All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
//       copyright notice, this list of conditions and the following
//       disclaimer in the documentation and/or other materials provided
//       with the distribution.
//     * Neither the name of Google Inc. nor the names of its
//       contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package io.flutter.plugins.fble;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothAdapter.LeScanCallback;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.content.Context;
import android.util.Pair;
import com.google.common.annotations.VisibleForTesting;
import com.google.protobuf.InvalidProtocolBufferException;
import io.flutter.plugins.fble.Protos.AdvertisementData;
import io.flutter.plugins.fble.Protos.GetLocalAdaptersResponse;
import io.flutter.plugins.fble.Protos.LocalAdapter;
import io.flutter.plugins.fble.Protos.ScanResult;
import io.flutter.plugins.fble.Protos.StartScanRequest;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.EventChannel.EventSink;
import io.flutter.plugin.common.EventChannel.StreamHandler;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import java.util.HashMap;
import java.util.UUID;

/**
 * Flutter Bluetooth Low Energy plugin.
 */
public class FblePlugin implements MethodCallHandler {

  private static final String GET_LOCAL_ADAPTERS = "getLocalAdapters";
  private static final String START_SCAN = "startScan";
  private static final String STOP_SCAN = "stopScan";
  private static final String NAMESPACE = "io.flutter.plugin.fble";
  private static final String METHOD_NAMESPACE = NAMESPACE + ".method";
  private static final String EVENT_NAMESPACE = NAMESPACE + ".event";

  private final Registrar registrar;
  private final AdvertisementParser parser;
  private final HashMap<String, Pair<EventChannel, AdapterSpecificScanCallback>> scanCallbacks;

  FblePlugin(Registrar registrar, AdvertisementParser parser) {
    this.registrar = registrar;
    this.parser = parser;
    scanCallbacks = new HashMap<>();
  }

  /**
   * Registers this native plugin with Flutter.
   *
   * @param registrar The Flutter plugin registrar.
   */
  public static void registerWith(Registrar registrar) {
    AdvertisementParser parser = new AdvertisementParser();
    FblePlugin myself = new FblePlugin(registrar, parser);
    MethodChannel methodChannel = new MethodChannel(registrar.messenger(), "fble");
    methodChannel.setMethodCallHandler(myself);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else if (GET_LOCAL_ADAPTERS.equals(call.method)) {
      getLocalAdapters(call, result);
    } else if (START_SCAN.equals(call.method)) {
      startScan(call, result);
    } else if (STOP_SCAN.equals(call.method)) {
      stopScan(call, result);
    } else {
      result.notImplemented();
    }
  }

  private BluetoothAdapter lookupAdapter(String _1) {
    // There is only one in Android.
    BluetoothManager bluetoothManager =
        (BluetoothManager) registrar.activity().getSystemService(Context.BLUETOOTH_SERVICE);
    return bluetoothManager.getAdapter();
  }

  @VisibleForTesting
  void getLocalAdapters(MethodCall _1, Result result) {
    // TODO: Request permissions and return empty list if not satisfied.
    BluetoothManager bluetoothManager =
        (BluetoothManager) registrar.activity().getSystemService(Context.BLUETOOTH_SERVICE);
    // Only one adapter in Android.
    BluetoothAdapter adapter = bluetoothManager.getAdapter();
    GetLocalAdaptersResponse proto = GetLocalAdaptersResponse.newBuilder()
        .setPlatform(GetLocalAdaptersResponse.Platform.ANDROID)
        .addAdapters(LocalAdapter.newBuilder().setOpaqueId(adapter.getAddress()))
        .build();
    result.success(proto.toByteArray());
  }

  @VisibleForTesting
  void startScan(MethodCall call, Result result) {
    // TODO: Request permission.
    byte[] data = call.arguments();
    StartScanRequest request = null;
    try {
      request = StartScanRequest.newBuilder().mergeFrom(data).build();
    } catch (InvalidProtocolBufferException e) {
      result.error("RuntimeException", e.getMessage(), e);
    }

    String adapterId = request.getAdapterId();
    BluetoothAdapter adapter = lookupAdapter(adapterId);
    EventChannel eventChannel = new EventChannel(registrar.messenger(),
        String.format("%s.scanResult.%s", EVENT_NAMESPACE, adapterId));

    AdapterSpecificScanCallback scanCallback;
    synchronized (scanCallbacks) {
      Pair<EventChannel, AdapterSpecificScanCallback> pair = scanCallbacks.get(adapterId);
      if (pair != null) {
        // Stop existing scan.
        adapter.stopLeScan(pair.second);
        pair.first.setStreamHandler(null);
      }
      scanCallback = new AdapterSpecificScanCallback(adapterId, parser);
      pair = new Pair<>(eventChannel, scanCallback);
      scanCallbacks.put(adapterId, pair);
    }

    eventChannel.setStreamHandler(scanCallback);

    UUID[] uuids = new UUID[request.getServiceUuidsCount()];
    int i = 0;
    for (String uuid : request.getServiceUuidsList()) {
      uuids[i++] = UUID.fromString(uuid);
    }
    if (uuids.length > 0) {
      adapter.startLeScan(uuids, scanCallback);
    } else {
      adapter.startLeScan(scanCallback);
    }
    result.success(null);
  }

  @VisibleForTesting
  void stopScan(MethodCall call, Result result) {
    String adapterId = call.arguments();
    Pair<EventChannel, AdapterSpecificScanCallback> pair;
    synchronized (scanCallbacks) {
      pair = scanCallbacks.get(adapterId);
      if (pair == null) {
        return;
      }
      scanCallbacks.remove(adapterId);
    }
    AdapterSpecificScanCallback scanCallback = pair.second;
    scanCallback.stop();
    pair.first.setStreamHandler(null);

    BluetoothAdapter adapter = lookupAdapter(adapterId);
    adapter.stopLeScan(scanCallback);
    result.success(null);
  }

  private static class AdapterSpecificScanCallback implements LeScanCallback, StreamHandler {

    // TODO: Use LeBluetoothScanner instead of this deprecated API.
    private final String id;
    private final AdvertisementParser parser;
    private EventSink eventSink;

    AdapterSpecificScanCallback(String id, AdvertisementParser parser) {
      this.id = id;
      this.parser = parser;
    }

    @Override
    public void onLeScan(BluetoothDevice device, int rssi, byte[] scanRecord) {
      AdvertisementData ads = parser.parse(scanRecord);
      ScanResult.Builder scanResult = ScanResult.newBuilder()
          .setRemoteId(device.getAddress())
          .setRssi(rssi)
          .setAdvertisementData(ads);
      // TODO: Fill in "connectable" on Android O and above.
      if (device.getName() != null) {
        scanResult.setName(device.getName());
      }
      if (eventSink != null) {
        eventSink.success(scanResult.build().toByteArray());
      }
    }

    @Override
    public void onListen(Object o, EventSink eventSink) {
      this.eventSink = eventSink;
    }

    @Override
    public void onCancel(Object o) {
      eventSink = null;
    }

    void stop() {
      if (eventSink != null) {
        eventSink.endOfStream();
      }
    }
  }
}

File: packages/fble/ios/Classes/FblePlugin.m
// Copyright 2017, the Flutter project authors. All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
//       copyright notice, this list of conditions and the following
//       disclaimer in the documentation and/or other materials provided
//       with the distribution.
//     * Neither the name of Google Inc. nor the names of its
//       contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import <CoreBluetooth/CoreBluetooth.h>

#import "Fble.pbobjc.h"
#import "FblePlugin.h"

static NSString *const kGetLocalAdapters = @"getLocalAdapters";
static NSString *const kStartScan = @"startScan";
static NSString *const kStopScan = @"stopScan";
static NSString *const kNamespace = @"io.flutter.plugins.fble";
static NSString *const kMethodNamespace = @"io.flutter.plugins.fble.method";
static NSString *const kEventNamespace = @"io.flutter.plugins.fble.event";
static NSString *const kDefaultAdapterId = @"02000000-0000-0000-0000-000000000000";

typedef void (^ScanResultCallback)(ProtosScanResult*);

@interface BluetoothAdapter : NSObject<CBCentralManagerDelegate> {
  ScanResultCallback _scanResultCallback;
}

@property (nonatomic, strong) CBCentralManager *centralManager;
@property (nonatomic, strong) NSString *identifier;
@property (nonatomic, strong, getter=getScanResultCallback, setter=setScanResultCallback:) ScanResultCallback scanResultCallback;

- (id)initWithIdentifier:(NSString*)identifier;
- (void)startScan:(NSArray<CBUUID *> *)services;
- (void)stopScan;
@end

@implementation BluetoothAdapter
- (id)initWithIdentifier:(NSString*)identifier {
  if ((self = [super init]) != nil) {
    self.centralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
    self.identifier = identifier;
    self->_scanResultCallback = nil;
    return self;
  } else {
    return nil;
  }
}

- (ScanResultCallback)getScanResultCallback {
  @synchronized (self) {
    return self->_scanResultCallback;
  }
}

- (void)setScanResultCallback:(ScanResultCallback) callback {
  @synchronized (self) {
    self->_scanResultCallback = callback;
  }
}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {
  // TODO: What?
}

- (void)centralManager:(CBCentralManager *)central didDiscoverPeripheral:(CBPeripheral *)peripheral advertisementData:(NSDictionary<NSString *,id> *)advertisementData RSSI:(NSNumber *)RSSI {
  if (!self->_scanResultCallback) {
    return;
  }
  ProtosScanResult *result = [[ProtosScanResult alloc] init];
  [result setName:[peripheral name]];
  [result setRemoteId:[[peripheral identifier] UUIDString]];
  [result setRssi:[RSSI intValue]];
  ProtosAdvertisementData *ads = [[ProtosAdvertisementData alloc] init];
  [ads setLocalName:advertisementData[CBAdvertisementDataLocalNameKey]];
  [ads setManufacturerData:advertisementData[CBAdvertisementDataManufacturerDataKey]];
  NSDictionary *serviceData = advertisementData[CBAdvertisementDataServiceDataKey];
  for (CBUUID *uuid in serviceData) {
    [[ads serviceData] setObject:serviceData[uuid] forKey:uuid.UUIDString];
  }
  [ads setTxPowerLevel:[advertisementData[CBAdvertisementDataTxPowerLevelKey] intValue]];
  [ads setConnectable:[advertisementData[CBAdvertisementDataIsConnectable] boolValue]];
  [result setAdvertisementData:ads];
  @synchronized (self) {
    // Double-checked.
    if (self->_scanResultCallback) {
      self->_scanResultCallback(result);
    }
  }
}

- (void)startScan:(NSArray<CBUUID *> *)services {
  NSDictionary *options = @{
      CBCentralManagerScanOptionAllowDuplicatesKey: @YES
  };
  if (services.count == 0) {
    services = nil;
  }
  [self.centralManager scanForPeripheralsWithServices:services options:options];
}

- (void)stopScan {
  [self.centralManager stopScan];
}
@end

@interface FblePlugin()
@property (nonatomic, strong) BluetoothAdapter *defaultAdapter;
@property (nonatomic, strong) NSObject<FlutterPluginRegistrar> *registrar;
@end

@implementation FblePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel = [FlutterMethodChannel
      methodChannelWithName:kMethodNamespace
            binaryMessenger:[registrar messenger]];
  FblePlugin *instance = [[FblePlugin alloc] initWithRegistrar:registrar];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)initWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  if ((self = [super init]) != nil) {
    self.registrar = registrar;
    self.defaultAdapter = [[BluetoothAdapter alloc] initWithIdentifier:kDefaultAdapterId];
    return self;
  } else {
    return nil;
  }
}

- (BluetoothAdapter *)lookupAdapter:(NSString *)identifier {
  return self.defaultAdapter;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([kGetLocalAdapters isEqualToString:call.method]) {
    [self getLocalAdapters:call result:result];
  } else if ([kStartScan isEqualToString:call.method]) {
    [self startScan:call result:result];
  } else if ([kStopScan isEqualToString:call.method]) {
    [self stopScan:call result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)getLocalAdapters:(FlutterMethodCall*)call result:(FlutterResult)result {
  ProtosGetLocalAdaptersResponse *response = [[ProtosGetLocalAdaptersResponse alloc] init];
  [response setPlatform:ProtosGetLocalAdaptersResponse_Platform_Ios];
  // There is no concept of separate adapters in iOS. So we return a fake one.
  ProtosLocalAdapter *defaultAdapter = [[ProtosLocalAdapter alloc] init];
  [defaultAdapter setOpaqueId:[self.defaultAdapter identifier]];
  [[response adaptersArray] addObject:defaultAdapter];
  result([FlutterStandardTypedData typedDataWithBytes:[response data]]);
}

- (void)startScan:(FlutterMethodCall*)call result:(FlutterResult)result {
  FlutterStandardTypedData *data = [call arguments];
  ProtosStartScanRequest *request = [[ProtosStartScanRequest alloc] initWithData:[data data] error:nil];
  NSString *adapterId = [request adapterId];
  BluetoothAdapter *adapter = [self lookupAdapter:adapterId];
  NSMutableArray<CBUUID *> *services = [NSMutableArray arrayWithCapacity:[request.serviceUuidsArray count]];
  for (NSString *uuidString in request.serviceUuidsArray) {
    CBUUID *uuid = [CBUUID UUIDWithString:uuidString];
    [services addObject:uuid];
  }
  [adapter startScan:services];
  FlutterEventChannel *bleEventChannel =
      [FlutterEventChannel eventChannelWithName:[kEventNamespace stringByAppendingFormat:@".scanResult.%@", adapterId]
                                binaryMessenger:[self.registrar messenger]];
  bleEventChannel.streamHandler = self; // TODO: Clean up the stream handler in onCancelWithArguments?
  result(nil);
}

- (void)stopScan:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString *adapterId = [call arguments];
  BluetoothAdapter *adapter = [self lookupAdapter:adapterId];
  [adapter stopScan];
  result(nil);
}

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)eventSink {
  self.defaultAdapter.scanResultCallback = ^(ProtosScanResult *result) {
    FlutterStandardTypedData *data = [FlutterStandardTypedData typedDataWithBytes:[[result data] copy]];
    eventSink(data);
  };
  return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
  self.defaultAdapter.scanResultCallback = nil;
  return nil;
}
@end
